### PR TITLE
Fix zero gradients for ppo-clipped vf

### LINKED ISSUE #19291: [Bug] PPO value function loss is incorrect
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I think the calculation of the clipped value function loss in PPO is incorrect, or at the very least the function of the "vf_clip_param" is misleading.

In the current implementation, `vf_loss1` is the error between the current value function output and the target. `vf_loss_clipped` is the difference between the previous value function and the current value function, clamped with a config clip parameter. `vf_loss2` is the error between `vf_loss_clipped` and the target. Then, the total value loss is computed as the mean of the MAX of `vf_loss1` and `vf_loss2`.

This doesn't make sense, because the clipping parameter really should be _clipping the value of the loss_. However, in the current implementation, the loss can be much bigger than the clipping parameter. For example, even if the clipping parameter is set to 0, then the loss is still either the difference between the current or the last value function output and the target (whichever one is larger). Note that when the clipping parameter is 0 the vf_clipped value simplifies to prev_value_function_out, and when it is inf the vf_clipped value simplifies to value_fn_out.

Let me provide a small example to demonstrate why this implementation doesn't make sense to me. Let's say that the target value is 1.0, and the last value function output was 0.0, but since the last iteration the model improved a bit so the value function output is currently 0.8. Let us also assume that our clipping parameter is 0.1. Then, we would calculate `vf_loss1 = (1.0 - 0.8)**2 = 0.04`,  `vf_loss_clipped = 0.0 + clamp(0.8 - 0.0, 0.1, -0.1) = 0.1`,  `vf_loss2 = (1.0 - 0.1)**2 = 0.81`,  `vf_loss = max(0.04, 0.81) = 0.81`. So, our loss is 0.81 even though we are very close to our target. This will cause us to overshoot. As you can see, the clip parameter did nothing to actually clamp the value function loss, and the max operation caused us to choose a loss that was far too large.

In contrast, I would expect the vf loss to be calculated by simply clamping `vf_loss1`:
`vf_loss = clamp( (value_target - curr_value_function_out)**2, clip_param, -clip_param)`

cc @smorad 

### Reproduction script

from `ppo_torch_policy.py:88`:
```python
prev_value_fn_out = train_batch[SampleBatch.VF_PREDS]
value_fn_out = model.value_function()
vf_loss1 = torch.pow(
            value_fn_out - train_batch[Postprocessing.VALUE_TARGETS], 2.0)
vf_clipped = prev_value_fn_out + torch.clamp(
            value_fn_out - prev_value_fn_out, -policy.config["vf_clip_param"],
            policy.config["vf_clip_param"])
vf_loss2 = torch.pow(
vf_clipped - train_batch[Postprocessing.VALUE_TARGETS], 2.0)
vf_loss = torch.max(vf_loss1, vf_loss2)
mean_vf_loss = reduce_mean_valid(vf_loss)
```

### Anything else

proposed updated code:
```python
prev_value_fn_out = train_batch[SampleBatch.VF_PREDS]
value_fn_out = model.value_function()
vf_loss1 = torch.pow(
            value_fn_out - train_batch[Postprocessing.VALUE_TARGETS], 2.0)
vf_loss = torch.clamp(
            vf_loss1, -policy.config["vf_clip_param"],
            policy.config["vf_clip_param"])
mean_vf_loss = reduce_mean_valid(vf_loss)
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by smorad**
## Why are these changes needed?
The PPO value loss calculation returns a zero-gradient when clipping is applied and `vf_loss2` is selected, because `prev_value_fn_out` is from the `SampleBatch` which doesn't track gradients. Furthermore, the logic itself is a bit convoluted. See the related issue for a more in-depth description.

## Related issue number

Closes #19291

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by smorad** [comment]
This results in much better using value losses. I modified the [tuned cartpole](https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/ppo/cartpole-ppo.yaml) example slightly to increase the value function los coefficient. By default, the value function coefficient is set very low in nearly all tuned PPO examples. I suspect this is because the value function is not behaving as expected, and setting it higher results in bad agents. 

With the following config:
```yaml
cartpole-ppo:
    env: CartPole-v0
    run: PPO
    stop:
        episode_reward_mean: 200
        timesteps_total: 100000
    config:
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 1
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_loss_coeff: 0.5
        model:
            fcnet_hiddens: [32]
            fcnet_activation: linear
            vf_share_layers: true
```

Screen Shot 2022-02-07 at 4 54 03 PM
Screen Shot 2022-02-07 at 4 53 33 PM

With the fix, we can see the value loss is much closer to what's expected (~7) rather than `master` (~1000). Note that this is the **mean value error**. That is, at a timestep t, we can expect the `master` value function to be off by 1000. This makes no sense for this environment where the cumulative reward is 100-200.

Second trial with a slightly different config:
```yaml
cartpole-ppo:
    env: CartPole-v0
    run: PPO
    stop:
        episode_reward_mean: 200
        timesteps_total: 100000
    config:
        # Works for both torch and tf.
        framework: torch
        gamma: 0.99
        lr: 0.0003
        num_workers: 1
        observation_filter: MeanStdFilter
        num_sgd_iter: 6
        vf_loss_coeff: 0.9
        model:
            fcnet_hiddens: [32]
            fcnet_activation: linear
            vf_share_layers: true
```
Screen Shot 2022-02-07 at 5 18 00 PM
Screen Shot 2022-02-07 at 5 12 28 PM

Cartpole is a simple environment that can be solved using vanilla policy gradient, which makes sense why the value function misbehaving has little effect on the final reward. I expect harder tasks that require a valid value function to solve would see a significant divergence between the final reward.

**Review by gjoliver — APPROVED** [review]
wow, this fixes the long outstanding value loss calculation?
thanks so much man!

**Comment by avnishn** [comment]
Is this the same case on the tf branch? Would be good to apply the fix there as well.

**Comment by smorad** [comment]
Tensorflow does indeed do the weird clipping as well:
https://github.com/ray-project/ray/blob/7f1bacc7dc9caf6d0ec042e39499bbf1d9a7d065/rllib/agents/ppo/ppo_tf_policy.py#L115-L129

**Comment by avnishn** [comment]
We'll need to make this change uniformly to both frameworks. Can you submit the change for tf as well? Thx 😊

**Comment by smorad** [comment]
Results after tf fix:
Screen Shot 2022-02-13 at 3 23 23 PM

**Review by sven1977 — APPROVED** [review]
Awesome! Thanks for the fixes @smorad !
